code/contracts/ONEWallet2.sol

```diff
@@ -0,0 +1,157 @@
+
+    /**
+     * This is a general commit and should works with all kinds of operations distinguished by otype, followed by operation-specific parameters. 
+     */
+    function commit(bytes32 hash, OperType otype, address payable addr, uint amount) external {
+        uint32 submittedOTPIdx = uint32(block.timestamp) / interval - t0;
+        require(submittedOTPIdx > lastExecutedIdx, "It is not possible to submit commits in the current OTP interval anymore. Wait for the next interval.");
+        Commit memory nc = Commit(hash, block.timestamp, otype, addr, amount);
+        commits[submittedOTPIdx].push(nc);                
+    }
+
+    /**
+     * This is a general reveal for all types of operation.
+     */
+    function reveal(bytes32[] calldata neighbors, uint32 otpIdx, bytes32 eotp) external  returns (bool) {                        
+        // 1) clean up
+        if(block.timestamp / 1 days - cleanupCheckpoint  >= AUTO_CLEANUP_DAYS) {
+            cleanUpDailySpendings();
+        }        
+
+        // 2) Check time constrains of reveal (it is fine also with recovery)
+        uint32 currentIdx = uint32(block.timestamp) / interval - t0; // check whether idx of OTP is timely
+        require(currentIdx - otpIdx <= MAX_REVEAL_DELAY, "Reveal is out of max allowed delay.");                                
+
+        // 3) scan all the commits made within otpIdx interval and execute the first matching one        
+        for (uint i = 0; i < commits[otpIdx].length; i++) {
+            Commit storage c = commits[otpIdx][i];
+            bytes32 hashedArgs = sha256(abi.encodePacked(eotp, c.operType, c.addr, c.amount));
+            if(hashedArgs == commits[otpIdx][i].hash){                
+                
+                // 4) verify the provided OTP in different way, according to the operation type
+                if(c.operType != OperType.RECOVERY){                                        
+                    _verifyMerkleProof(neighbors, otpIdx, eotp);
+                } else {
+                    require(sha256(abi.encodePacked(eotp)) == recoveryHash, "Provided secret for recovery is incorrect.");
+                }
+
+                // 5) execute the operation according to its type
+                if(c.operType == OperType.TRANSFER){
+                    _executeTransfer(c.addr, c.amount);                
+                } else if(c.operType == OperType.SET_LAST_RESORT_ADDR){
+                    if(lastResortAddress == address(0x0) && c.addr != address(0x0))
+                        lastResortAddress = c.addr;
+                } else if(c.operType == OperType.RECOVERY){ // we do not care about params in Commits
+                    _drain(); // IH: I do not find any reason for this operation to be executed using OTP.
+                } else {
+                    emit UnknownOperationType(c.operType);                    
+                }                
+                
+                delete commits[otpIdx][i];
```

Comment:
This and the next line should be above the execution block (line 90). So even if the execution becomes out-of-gas , the commit can be deleted and lastExecutedIdx is properly set to new index.

Reply:
I do not agree. All transactions are **atomic**, so they are either fully executed or not executed at all. If there is out-of-gas error, everything is just reverted to its original state.

Reply:
Yes, that's true. Maybe I was half asleep when I wrote that :) I forgot the context. Anyway, this issue already becomes moot in the new version.